File: 0401-0500/0437-Path Sum III/0437-Path Sum III.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root is None:
            return 0
        
        def dfs(root, sum):
            if root is None:
                return 0
            sum -= root.val
            count = 1 if sum == 0 else 0
            return dfs(root.left, sum) + dfs(root.right, sum) + count
        return self.pathSum(root.left, sum) + self.pathSum(root.right, sum) + dfs(root, sum)

import collections
class Solution2:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        table = collections.Counter([0])
        count = 0
        def dfs(root, curr):
            nonlocal count
            if root is None:
                return
            curr += root.val
            count += table[curr - sum]
            table[curr] += 1
            dfs(root.left, curr)
            dfs(root.right, curr)
            table[curr] -= 1
        dfs(root, 0)    
        return count

File: 0401-0500/0437-Path Sum III/0437-Path Sum III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int sum) {
        if (root == nullptr) {
            return 0;
        }
        
        int leftNum = pathSum(root->left, sum);
        int rightNum = pathSum(root->right, sum);
        
        int count = 0;
        dfs(root, sum, count);
        return leftNum + rightNum + count;
    }
private:
    void dfs(TreeNode* root, int sum, int& count) {
        if (root == nullptr) {
            return;
        }
        
        sum -= root->val;
        if (sum == 0) {
            ++count;
        }
        dfs(root->left, sum, count);
        dfs(root->right, sum, count);
    }
};

class Solution2 {
public:
    int pathSum(TreeNode* root, int sum) {
        table[0] = 1;
        int count = 0;
        dfs(root, 0, sum, count);
        return count;
    }
    
private:
    unordered_map<int, int> table;
    void dfs(TreeNode* root, int curr, int sum, int& count) {
        if (root == nullptr) {
            return;
        }
        
        curr += root->val;
        count += table[curr - sum];
        ++table[curr];
        dfs(root->left, curr, sum, count);
        dfs(root->right, curr, sum, count);
        --table[curr];
    }
};
